Add blog article about volume expansion graduating to GA (#32663)
kubernetes/website: content/en/blog/_posts/2022-04-18-volume-expansion-ga.md

---
layout: blog
title: "Volume expansion now available as stable feature in Kubernetes 1.24"

> date: 2022-04-18

Reviewer: ```suggestion
date: 2022-05-05
```

---
layout: blog

> title: "Volume expansion going GA in v1.24"

Author: Changed the title.

date: 2022-04-18
slug: volume-expansion-ga
---

> **Author:** Hemant Kumar

Reviewer 2: Is this right?
```suggestion
**Author:** Hemant Kumar (Red Hat)
```

> ## Introduction

Reviewer 2: ```suggestion
```
(We don't start articles with headings)

Volume expansion was introduced as a alpha feature in Kubernetes 1.8 and it went beta in 1.11 and with Kubernetes 1.24 we are excited to announce general availability(GA)
of volume expansion feature.

This feature allows Kubernetes users to simply edit their `PersistentVolumeClaim` objects and specify new size in PVC Spec and Kubernetes will automatically expand the volume
using storage backend and also expand the underlying file system in-use by the Pod without requiring any downtime at all if possible.

### How to use volume expansion feature

> Using volume expansion is as simple as editing `storage` field of pvc spec and specifying new size. For example - given following PVC:

Author: heh - that implies `storage` is a direct child inside `spec`, but it is more like `.spec.resources.requests.storage` is the one you want to change. I think that full "path" of storage field is too long to be used in a blog like article but just specifying `.spec.storage` is basically incorrect.

> Using volume expansion is as simple as editing `storage` field of pvc spec and specifying new size. For example - given following PVC:

Reviewer 2: TBH I didn't check the API reference _or_ the example just below. Oops.

I edited the suggestion.

```
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: myclaim
spec:
  accessModes:
    - ReadWriteOnce
  resources:
    requests:

>       storage: 1Gi --> specify new size here

Author: Saying "sizes are in bytes" is not entirely accurate because as an API field in JSON (at least the one user is editing) it is basically a *string*.  Sure it can be bytes but we also try and best to convert string values like `1Gi` or `10GB` to appropriate byte sizes internally. 

I am not sure if a blog article is right place to inform users about nuances associated with `resource.Quantity` type and how it is serialized internally.

>       storage: 1Gi --> specify new size here

Reviewer 2: Serialization and any Golang implementation details aren't important, but folks reading this article might well not know that Kubernetes has implied units (`storage: 1Gi` not `storageBytes: 1Gi`, nor `storage: 1GiB`). I have definitely lost count of the number of production Kubernetes clusters I've seen where key workloads share nodes with Pods that don't have requests and limits...

We should aim to cover things for those less experienced readers. If people have to fire up a Google search to understand what they're reading here, I don't think we've succeeded.

You can link to https://kubernetes.io/docs/reference/glossary/?all=true#term-quantity if that helps.

```

Users can request expansion of the underlying volume by specifying a new value instead of old `1Gi` size. Once expansion is initiated - pvc's conditions can be monitored
for completion of volume expansion operation.

When Kubernetes starts expanding the volume - it will add `Resizing` condition to the PVC, which will be removed once expansion completes. More information about progress of
expansion operation can also be obtained by monitoring events associated with pvc:

```

> ~> kubectl describe pvc <pvc>

Reviewer 2: ```suggestion
kubectl describe pvc <pvc>
```
(also, this isn't actually a watch - maybe explain how to watch for the events as well?)

```

### Storage driver support

Not every volume type however is expandable by default. Some volume types such as - intree hostpath volumes are not expandable at all. For CSI volumes - the CSI driver
must have capability `EXPAND_VOLUME` in controller or node service (or both if appropriate). For CSI volumes - please refer to documentation of your CSI driver, to find out
if it supports volume expansion.

Please refer to volume expansion documentation for intree volume types which support volume expansion - [Expanding Persistent Volumes](/docs/concepts/storage/persistent-volumes/#expanding-persistent-volumes-claims)

In general to provide some degree of control over volumes that can be expanded, only dynamically provisioned PVCs whose storage class has `allowVolumeExpansion` parameter set to `true` are expandable.

A Kubernetes admin must edit storage class object and set `allowVolumeExpansion` field to true. For example:

```
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:

>   name: gp2-default

Reviewer 2: (nit)
```suggestion
  name: gp3-default
```
**or** use a fictional provisioner, eg `cloud.example/cloud-block-storage`

Reviewer 2: In light of https://github.com/kubernetes/website/issues/33036, let's _not_ suggest `gp3` as the example.

provisioner: kubernetes.io/aws-ebs
parameters:
  secretNamespace: ""
  secretName: ""
allowVolumeExpansion: true
```

### Online vs offline expansion

> By default Kubernetes attempts to expand volumes immediately after user requests new size. Expansion happens online if one or more Pods are using the volume and as a result volume expansion requires no application downtime. File System expansion on the node is also performed online and hence does not require shutting down the Pod that was using the PVC.

Reviewer 2: ```suggestion
By default, Kubernetes attempts to expand volumes immediately after user requests a resize.
If one or more Pods are using the volume, Kubernetes tries to expands the volume using an online resize;
as a result volume expansion usually requires no application downtime.
Filesystem expansion on the node is also performed online and hence does not require shutting
down any Pod that was using the PVC.

If you expand a PersistentVolume that is not in use, Kubernetes does an offline resize (and,
because the volume isn't in use, there is again no workload disruption).
```
?

In some cases though - if underlying Storage Driver can only support offline expansion, users of the PVC must take down their Pod before expansion can succeed. Please refer to documentation of your storage
provider to find out - what mode of volume expansion it supports.

> When volume expansion was introduced as an alpha feature, it only supported offline file system expansion on the node and hence required users to restart their pods for file system resizing to finish. This behaviour has been changed and Kubernetes tries its best to full-fill user's request regardless of the fact that if volume is online or offline. If Storage Provider supports online expansion - no Pod restart should be necessary for volume expansion to finish.

Reviewer 2: ```suggestion
When volume expansion was introduced as an alpha feature, Kubernetes only supported offline filesystem
expansion on the node and hence required users to restart their pods for file system resizing to finish. 
his behaviour has been changed and Kubernetes tries its best to fulfil any resize request regardless
of whether the underlying PersistentVolume volume is online or offline. If your storage provider supports
online expansion then no Pod restart should be necessary for volume expansion to finish.
```
Should “restart” actually be “replacement”?

> ### ## What’s next?

Reviewer 2: ```suggestion
## Next steps
```
“What's next” is a term we give a specific meaning to in the docs.

> Although Volume expansion is going GA with 1.24 - we are continously working to make it simpler for users of Kubernetes to expand their persistent volumes. Kubernetes 1.23 introduced feature `RecoverVolumeExpansionFailure` - so as users can themselves recover from volume expansion failures (usually recovering from volume expansion requires admin intervention) whenever possible. See - [Rcovering from volume expansion failure](/docs/concepts/storage/persistent-volumes/#recovering-from-failure-when-expanding-volumes) for more details.

Reviewer 2: ```suggestion
Although volume expansion is now stable as part of the recent v1.24 release,
SIG Storage are working to make it even simpler for users of Kubernetes to expand their persistent storage.
Kubernetes 1.23 introduced features for triggering recovery from failed volume expansion, allowing users
to attempt self-service healing after a failed resize.
See [Recovering from volume expansion failure](/docs/concepts/storage/persistent-volumes/#recovering-from-failure-when-expanding-volumes) for more details.
```

> Kubernetes community is also working on statefulset expansion - so as all underlying PVCs created by a statefulset can be expanded by directly editing statefulset template. See - [Statefulset volume resize](https://github.com/kubernetes/enhancements/pull/2842) feature for more details.

Reviewer 2: ```suggestion
The Kubernetes contributor community is also discussing the potential for StatefulSet-driven storage expansion. This proposed
feature would let you trigger expansion for all underlying PVs that are providing storage to a StatefulSet,
by directly editing the StatefulSet object.
See the [Support Volume Expansion Through StatefulSets](https://github.com/kubernetes/enhancements/issues/661) enhancement proposal for more details.
```
